### ISSUE TITLE
Latest code throws 'unexpected keyword argument parent_page' in django-modelcluster

### ISSUE THREAD

**Issue by fozcode** [torchbox/wagtail#2267]
OK I probably shouldn't be using the bleeding edge code, but I need some of the 1.4 fixes and features for my project and it was working well before.  I'm getting an error with the latest code ever since the commits on Jan 25th to add the new permissions policy.  Every time I try to create a new page I get:

```
Traceback (most recent call last):
  File "/Users/matt/Dev/MyProject/MyApp/lib/python3.4/site-packages/django/core/handlers/base.py", line 149, in get_response
    response = self.process_exception_by_middleware(e, request)
  File "/Users/matt/Dev/MyProject/MyApp/lib/python3.4/site-packages/django/core/handlers/base.py", line 147, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/Users/matt/Dev/MyProject/MyApp/lib/python3.4/site-packages/django/views/decorators/cache.py", line 43, in _cache_controlled
    response = viewfunc(request, *args, **kw)
  File "/Users/matt/Dev/MyProject/MyApp/lib/python3.4/site-packages/django/contrib/auth/decorators.py", line 23, in _wrapped_view
    return view_func(request, *args, **kwargs)
  File "/Users/matt/Dev/MyProject/MyApp/MyApp/wagtail/wagtailadmin/views/pages.py", line 169, in create
    parent_page=parent_page)
  File "/Users/matt/Dev/MyProject/MyApp/lib/python3.4/site-packages/modelcluster/forms.py", line 208, in __init__
    super(ClusterForm, self).__init__(data, files, instance=instance, prefix=prefix, **kwargs)
TypeError: __init__() got an unexpected keyword argument 'parent_page'
```

The old code used to call `validate_page_form(form, parent_page)` but this has been deleted and replaced by `form = form_class(request.POST, request.FILES, instance=page, parent_page=parent_page)` along with the offending `parent_page`.

This looks like an accidental bug (possibly triggered by my page models?) but I don't understand this part of the code well enough to know how to fix it.  Any help appreciated, even if it's "don't use the latest code yet".

Thanks!

**Comment by gasman** [comment]
Hi @fozcode,
I wouldn't discourage anyone from using the bleeding edge code as long as they're prepared to encounter the occasional bit of instability - the more testing it gets, the better :-)

Afraid I can't see how this would be occurring, unless you've set a custom base_form_class on your page model (which is a Wagtail-1.4-specific customisation that's as yet undocumented). Assuming you haven't done that, get_form_class would return a form based on wagtail.wagtailadmin.forms.WagtailAdminModelForm (as set here https://github.com/torchbox/wagtail/blob/417cafe69bd2ae19c85790fcbc938d190e701165/wagtail/wagtailadmin/edit_handlers.py#L714), which should be filtering out the parent_page argument (https://github.com/torchbox/wagtail/blob/417cafe69bd2ae19c85790fcbc938d190e701165/wagtail/wagtailadmin/forms.py#L275). Since that's not happening, it looks like get_form_class is returning a form class that's not the one we expect - could you perhaps add some debug code to wagtail/wagtailadmin/views/pages.py so that we can see what form class is being used there?

**Comment by fozcode** [comment]
Hi @gasman,

Thanks for taking the time to look at this.  I haven't used a custom base_form_class, but here's what's happening:

In `wagtailadmin/views/pages.py` we have the new code:

    page = page_class(owner=request.user)
    edit_handler_class = page_class.get_edit_handler()
    form_class = edit_handler_class.get_form_class(page_class)

In my case, `page_class` is `<class 'bulletin_core.models.Submission'>`, which is the model I am trying to create, so I think this is correct.
Then `edit_handler_class` comes back as `<class 'wagtail.wagtailadmin.edit_handlers._TabbedInterface'>`, which seems OK from what I can tell.
Then `form_class` comes back as `<class 'django.forms.widgets.SubmissionForm'>`, which is a non-existent class and is clearly wrong.

To cut a long debugging story a bit shorter, it turns out that in my case the original import of `WagtailAdminModelForm` at: https://github.com/torchbox/wagtail/blob/master/wagtail/wagtailadmin/edit_handlers.py#L24
is somehow coming back with the mysterious `django.forms.widgets.WagtailAdminModelForm`.
If I look at the neighbouring value of `WagtailAdminPageForm` it is the correct `wagtail.wagtailadmin.forms.WagtailAdminPageForm`.

It would appear that for this one class, the `from .forms import X` is actually finding `django.forms` instead of `wagtailadmin.forms`, or something like that.  The class name it's finding doesn't exist in my django install (according to grep) so I wonder if this is a Python bug or maybe I have some kind of memory corruption going on.  Let me know if you can think of a valid explanation though!

**Comment by fozcode** [comment]
A bit more digging has led me to the original definition of `WagtailAdminModelForm` at:
https://github.com/torchbox/wagtail/blob/master/wagtail/wagtailadmin/forms.py#L262

    WagtailAdminModelForm = WagtailAdminModelFormMetaclass(str('WagtailAdminModelForm'), (ClusterForm,), {})

Now if you follow the hierarchy of `ClusterForm` all the way back, you end up in `django.forms.widgets`. So is the `django.forms.widgets.WagtailAdminModelForm` class name actually correct?

**Comment by fozcode** [comment]
In fact at:
https://github.com/torchbox/wagtail/blob/417cafe69bd2ae19c85790fcbc938d190e701165/wagtail/wagtailadmin/edit_handlers.py#L273

If I change: ```BaseFormEditHandler.base_form_class = WagtailAdminModelForm```

to be: ```BaseFormEditHandler.base_form_class = WagtailAdminPageForm```

then my errors go away.  There's another instance of `WagtailAdminModelForm` at:
https://github.com/torchbox/wagtail/blob/417cafe69bd2ae19c85790fcbc938d190e701165/wagtail/wagtailadmin/edit_handlers.py#L33
that I'm not sure about.

**Comment by fozcode** [comment]
To elaborate on that last comment, here is the model attribute that causes the error to be thrown:

    # Override the tab names
    edit_handler = TabbedInterface([
        ObjectList(content_panels, heading='Tab 1'),
        ObjectList(settings_panels, heading='Tab 2', classname='settings'),
    ])

I no longer need to override the tab names in my case, so I've removed this model attribute and the latest code now works.

The technique of customising the tabs is documented at:
http://docs.wagtail.io/en/v1.3.1/advanced_topics/customisation/page_editing_interface.html

**Comment by gasman** [comment]
Closing as unable to replicate.

**Comment by diveru4i** [comment]
Wagtail 1.4.1
Same problem here. This issue occurs if set edit_handler to my pages. Even if it is empty
```python
    edit_handler = TabbedInterface([])
```

**Comment by ababic** [comment]
We're having similar issues when upgrading a site to 1.4

**Comment by shredding** [comment]
Funny that we all encounter the same problem within minutes. I have the same problem.

I have a working fix, though:

```
from wagtail.wagtailadmin.edit_handlers import TabbedInterface as OriginalTabbedInterface

class TabbedInterface(OriginalTabbedInterface):

    def __init__(self, children, base_form_class=WagtailAdminPageForm):
        super().__init__(children, base_form_class)
```

-> Use this TabbedInterface instead of the shipped one.

@gasman: can you confirm that this is sufficient or do you see any side effects? AFAIK only the base_form_class must be adjusted.

**Comment by iaindillingham** [comment]
@shredding Thanks for the fix. I ran into a similar problem, followed your suggestion, and it works. I haven't gone into the details, but the problem seems to originate with ``TabbedInterface``.

**Comment by mihow** [comment]
Thanks @shredding, your fix worked for me as well, although I had to adjust it for Python 2.7
Python 3: `super()`
Python 2.7: `super(TabbedInterface, self) `

**Comment by dszmaj** [comment]
works for me either :) thanks @shredding

**Comment by m1kola** [comment]
I see some ways how to fix this:

1. We can revert some parts of 92d507752ff818801bafeb804ac7c25c2f2f89ea. In this case we probably can lost ability to easily override form for edit handlers. If I not mistaken this feature is not documented yet. I don't think that this is the best way to fix this bug, but this probably will be painless way if we want to fix this issue in 1.4.* branch. In 1.5 we can try to implement this in another way.

2. We can define `BaseFormEditHandler.base_form_classes` map instead of [`BaseFormEditHandler.base_form_class`](https://github.com/torchbox/wagtail/blob/v1.4.1/wagtail/wagtailadmin/edit_handlers.py#L273) for Pages and other models. This will be something like Django REST framework's [`HTMLFormRenderer.default_style`](https://github.com/tomchristie/django-rest-framework/blob/3.3.3/rest_framework/renderers.py#L256). For implementor's models we can provide hook and\or argument in constructor (argument already implemented in [`TabbedInterface`](https://github.com/torchbox/wagtail/blob/v1.4.1/wagtail/wagtailadmin/edit_handlers.py#L298) and in [`ObjectList`](https://github.com/torchbox/wagtail/blob/v1.4.1/wagtail/wagtailadmin/edit_handlers.py#L317)). Example of `BaseFormEditHandler`:

    ```python
    from wagtail.wagtailcore.models import Page
    from django.db import models

    # ...

    class BaseFormEditHandler(BaseCompositeEditHandler):
        # ...

        base_form_classes = ClassLookupDict({
            Page: WagtailAdminPageForm,
            models.Model: WagtailAdminModelForm,
        })

        # ...
    ```

3. We can just replace `WagtailAdminModelForm` with `WagtailAdminPageForm` in [`wagtail.wagtailadmin.edit_handlers`](https://github.com/torchbox/wagtail/blob/v1.4.1/wagtail/wagtailadmin/edit_handlers.py#L273) (probably we will need to do this in some other places too). Yes, this will work, but we have not only Pages, but Snippets. Current version of `WagtailAdminPageForm` will also work with Snippets, but *I think this is potentially dangerous and semantically wrong way to fix issue*.

4. We also can introduce `TabbedInterface` and `ObjectList` for Snippets... In my opinion this is dumb way.

@gasman and @timheap, what do you think? I can fix this bug, but we need to choose an approach. I vote for second way, but I'm not sure that we need to use this approach in 1.4.*

**Comment by timheap** [comment]
I've submitted a PR that should fix this issue. When adding a custom EditHandler to a Page class without setting a custom base_form_class for the EditHandler, the base_form_class is now taken from the model instead. This should make all previous custom EditHandlers work again.

I've also added a system check to ensure that form classes returned from Page EditHandlers subclass WagtailAdminPageForm, which should warn people when they make a custom EditHandler that would break the Page edit views.
